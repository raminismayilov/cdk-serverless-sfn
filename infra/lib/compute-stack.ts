import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';
import {
    aws_lambda as lambda,
    aws_stepfunctions as sfn,
    aws_stepfunctions_tasks as tasks,
    Duration,
} from "aws-cdk-lib";
import {
    NodejsFunction
} from "aws-cdk-lib/aws-lambda-nodejs";
import path from "path";

interface ComputeStackProps extends cdk.StackProps {
    stageName: string;
}

export class ComputeStack extends cdk.Stack {
    constructor(scope: Construct, id: string, props: ComputeStackProps) {
        super(scope, id, props);

        const addition = new NodejsFunction(this, 'Addition', {
            entry: path.join(__dirname, '..', '..', 'app', 'addition', 'addition.ts'),
            runtime: lambda.Runtime.NODEJS_16_X,
            functionName: `${props?.stageName}-addition`,
        });

        const square = new NodejsFunction(this, 'Square', {
            entry: path.join(__dirname, '..', '..', 'app', 'square', 'square.ts'),
            runtime: lambda.Runtime.NODEJS_16_X,
            functionName: `${props?.stageName}-square`,
        });

        const multiplication = new NodejsFunction(this, 'Multiplication', {
            entry: path.join(__dirname, '..', '..', 'app', 'multiplication', 'multiplication.ts'),
            runtime: lambda.Runtime.NODEJS_16_X,
            functionName: `${props?.stageName}-multiplication`,
        });

        const additionStep = new tasks.LambdaInvoke(this, 'Addition Step', {
            lambdaFunction: addition,
            inputPath: '$',
            outputPath: '$',
        });

        const squareStep = new tasks.LambdaInvoke(this, 'Square Step', {
            lambdaFunction: square,
            inputPath: '$.Payload',
            outputPath: '$',
        });

        const waitStep = new sfn.Wait(this, 'Wait Step', {
            time: sfn.WaitTime.duration(Duration.seconds(3)),
        });

        const definition = additionStep.next(waitStep).next(squareStep);

        const simpleStateMachine = new sfn.StateMachine(this, 'Simple State Machine', {
            definition,
            timeout: Duration.minutes(5),
            stateMachineName: `${props?.stageName}-simple-state-machine`,
        });
    }
}
