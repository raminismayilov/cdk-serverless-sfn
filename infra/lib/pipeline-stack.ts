import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { Artifact, IStage, Pipeline } from "aws-cdk-lib/aws-codepipeline";
import {
    CloudFormationCreateUpdateStackAction,
    CodeBuildAction,
    CodeBuildActionType,
    GitHubSourceAction,
} from "aws-cdk-lib/aws-codepipeline-actions";
import {
    BuildEnvironmentVariableType,
    BuildSpec,
    LinuxBuildImage,
    PipelineProject,
} from "aws-cdk-lib/aws-codebuild";

export class PipelineStack extends cdk.Stack {
    constructor(scope: Construct, id: string, props?: cdk.StackProps) {
        super(scope, id, props);

        const pipeline = new Pipeline(this, 'Pipeline', {
            pipelineName: 'MyPipeline',
            crossAccountKeys: false,
        });

        const sourceOutput = new Artifact('SourceOutput');

        const sourceAction = new GitHubSourceAction({
            actionName: 'PipelineSource',
            owner: 'raminismayilov',
            repo: 'cdk-serverless-sfn',
            branch: 'master',
            oauthToken: cdk.SecretValue.secretsManager('github-token'),
            output: sourceOutput,
        });

        pipeline.addStage({
            stageName: 'Source',
            actions: [sourceAction],
        });

        const buildOutput = new Artifact('BuildOutput');

        const buildAction = new CodeBuildAction({
            actionName: "PipelineBuild",
            input: sourceOutput,
            outputs: [buildOutput],
            project: new PipelineProject(this, "BuildProject", {
                environment: {
                    buildImage: LinuxBuildImage.STANDARD_5_0,
                },
                buildSpec: BuildSpec.fromSourceFilename(
                    "infra/build-specs/build-spec.yml"
                ),
            }),
        });

        pipeline.addStage({
            stageName: "Build",
            actions: [
                buildAction,
            ],
        });
    }
}
