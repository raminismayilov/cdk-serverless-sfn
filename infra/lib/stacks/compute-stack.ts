import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';
import {
    aws_lambda as lambda,
    aws_stepfunctions as sfn,
    aws_stepfunctions_tasks as tasks,
    CfnOutput,
    Duration,
} from 'aws-cdk-lib';
import { NodejsFunction } from 'aws-cdk-lib/aws-lambda-nodejs';
import path from 'path';
import { ApplicationAPI } from "../api";

interface ComputeStackProps extends cdk.StackProps {
    stageName: string;
}

export class ComputeStack extends cdk.Stack {
    public readonly multiplicationApiUrl: CfnOutput;

    constructor(scope: Construct, id: string, props: ComputeStackProps) {
        super(scope, id, props);

        const addition = new NodejsFunction(this, 'Addition', {
            entry: path.join(__dirname, '..', '..', '..', 'app', 'addition', 'addition.ts'),
            runtime: lambda.Runtime.NODEJS_16_X,
            functionName: `${props?.stageName}-addition`,
        });

        const square = new NodejsFunction(this, 'Square', {
            entry: path.join(__dirname, '..', '..', '..', 'app', 'square', 'square.ts'),
            runtime: lambda.Runtime.NODEJS_16_X,
            functionName: `${props?.stageName}-square`,
        });

        const multiplication = new NodejsFunction(this, 'Multiplication', {
            entry: path.join(__dirname, '..', '..', '..', 'app', 'multiplication', 'multiplication.ts'),
            runtime: lambda.Runtime.NODEJS_16_X,
            functionName: `${props?.stageName}-multiplication`,
        });

        const applicationApi = new ApplicationAPI(this, 'ApplicationAPI', {
            multiplicationService: multiplication,
        });
        this.multiplicationApiUrl = applicationApi.multiplicationApiUrl;

        const additionStep = new tasks.LambdaInvoke(this, 'Addition Step', {
            lambdaFunction: addition,
            inputPath: '$',
            outputPath: '$',
        });

        const squareStep = new tasks.LambdaInvoke(this, 'Square Step', {
            lambdaFunction: square,
            inputPath: '$.Payload',
            outputPath: '$',
        });

        const waitStep = new sfn.Wait(this, 'Wait Step', {
            time: sfn.WaitTime.duration(Duration.seconds(3)),
        });

        const definition = additionStep.next(waitStep).next(squareStep);

        const simpleStateMachine = new sfn.StateMachine(this, 'Simple State Machine', {
            definition,
            timeout: Duration.minutes(5),
            stateMachineName: `${props?.stageName}-simple-state-machine`,
        });
    }
}
