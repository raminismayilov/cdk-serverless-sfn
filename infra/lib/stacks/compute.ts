import { CfnOutput, Stack, StackProps } from 'aws-cdk-lib';
import { aws_ec2 as ec2, aws_iam as iam, aws_rds as rds } from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { Services } from '../services';
import { API } from '../api';

interface ComputeStackProps extends StackProps {
    vpc: ec2.IVpc;
    privateSg: ec2.SecurityGroup;
    rdsCredentials: rds.DatabaseSecret;
    rdsProxy: rds.DatabaseProxy;
}

export class ComputeStack extends Stack {
    public readonly api: API;

    constructor(scope: Construct, id: string, props: ComputeStackProps) {
        super(scope, id, props);

        const services = new Services(this, 'Services', props);

        this.api = new API(this, 'Pv', {
            moduleService: services.moduleService,
            migrationService: services.migrationService,
        });
    }
}
