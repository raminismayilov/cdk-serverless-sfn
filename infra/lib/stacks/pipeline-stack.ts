import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';
import {
    BuildSpec, EventAction, FilterGroup, GitHubSourceCredentials, Project, Source
} from 'aws-cdk-lib/aws-codebuild';
import { CodeBuildStep, CodePipeline, CodePipelineSource } from "aws-cdk-lib/pipelines";
import { Deployment } from "../stages/deployment";
import { pipelines } from "aws-cdk-lib";

export class PipelineStack extends cdk.Stack {
    private readonly pipeline: CodePipeline;

    constructor(scope: Construct, id: string, props?: cdk.StackProps) {
        super(scope, id, props);

        const owner = 'raminismayilov';
        const repo = 'cdk-serverless-sfn';
        const branch = 'master';
        const token = cdk.SecretValue.secretsManager('github-token');

        new GitHubSourceCredentials(this, 'GitHubCreds', {
            accessToken: token,
        });

        const buildSpec = BuildSpec.fromObject({
            version: 0.2,
            phases: {
                install: {
                    commands: ['n 16', 'node -v', 'npm ci'],
                },
                build: {
                    commands: ['npm run test:infra', 'npm run cdk:synth']
                },
            },
        });

        const synthAction = new CodeBuildStep(`Synth`, {
            input: CodePipelineSource.gitHub(`${owner}/${repo}`, branch, {
                authentication: token,
            }),
            partialBuildSpec: buildSpec,
            commands: [],
            primaryOutputDirectory: "infra/cdk.out",
        });

        this.pipeline = new CodePipeline(this, 'Pipeline', {
            synth: synthAction,
        });

        const testStage = new Deployment(this, 'Test', {
            env: props?.env
        });

        this.pipeline.addStage(testStage, {
            post: [
                new CodeBuildStep('Test', {
                    commands: ['n 16', 'node -v', 'npm ci', 'npm run test:app'],
                    envFromCfnOutputs: {
                        MULTIPLICATION_API_URL: testStage.multiplicationApiUrl,
                    }
                })
            ]
        });

        const prodStage = new Deployment(this, 'Prod', {
            env: props?.env
        });

        this.pipeline.addStage(prodStage, {
            pre: [
                new pipelines.ManualApprovalStep('PromoteToProd'),
            ]
        });
    }
}
