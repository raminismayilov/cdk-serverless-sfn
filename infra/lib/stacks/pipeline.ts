import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';
import {
    BuildSpec,
    EventAction,
    FilterGroup,
    GitHubSourceCredentials,
    Project,
    Source,
} from 'aws-cdk-lib/aws-codebuild';
import {
    CodeBuildStep,
    CodePipeline,
    CodePipelineSource,
    ManualApprovalStep,
    ShellStep,
} from 'aws-cdk-lib/pipelines';
import { pipelines, Stage } from 'aws-cdk-lib';
import { TestStage } from '../stages/test-stage';

export class PipelineStack extends cdk.Stack {
    private readonly pipeline: CodePipeline;

    constructor(scope: Construct, id: string, props: cdk.StackProps) {
        super(scope, id, props);

        const owner = 'raminismayilov';
        const repo = 'cdk-serverless-sfn';
        const branch = 'master';
        const token = cdk.SecretValue.secretsManager('github-token');

        new GitHubSourceCredentials(this, 'GitHubCreds', {
            accessToken: token,
        });

        const buildSpec = BuildSpec.fromObject({
            version: 0.2,
            phases: {
                install: {
                    commands: ['n 16', 'node -v', 'npm ci'],
                },
                build: {
                    commands: ['npm run test:infra', 'npm run cdk:synth'],
                },
            },
        });

        const synthAction = new CodeBuildStep(`Synth`, {
            input: CodePipelineSource.gitHub(`${owner}/${repo}`, branch, {
                authentication: token,
            }),
            partialBuildSpec: buildSpec,
            commands: [],
            primaryOutputDirectory: 'infra/cdk.out',
        });

        this.pipeline = new CodePipeline(this, 'Pipeline', {
            synth: synthAction,
        });

        const testStage = new TestStage(this, 'TestStage', { env: props.env });

        this.pipeline.addStage(testStage, {
            post: [
                new CodeBuildStep('Test', {
                    commands: ['n 16', 'node -v', 'npm ci', 'npm run test:app'],
                    envFromCfnOutputs: {
                        API_URL: testStage.apiUrl,
                    },
                }),
            ],
        });
    }
}
