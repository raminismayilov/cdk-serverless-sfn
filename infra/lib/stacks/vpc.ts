import { Construct } from 'constructs';
import { aws_ec2 as ec2, aws_iam as iam, Stack } from 'aws-cdk-lib';

export class VpcStack extends Stack {
    public readonly vpc: ec2.IVpc;

    constructor(scope: Construct, id: string) {
        super(scope, id);

        // VPC with public and private subnets
        this.vpc = new ec2.Vpc(this, 'serverless-app', {
            cidr: '10.0.0.0/20',
            natGateways: 0,
            maxAzs: 2,
            enableDnsHostnames: true,
            enableDnsSupport: true,
            subnetConfiguration: [
                {
                    cidrMask: 22,
                    name: 'public',
                    subnetType: ec2.SubnetType.PUBLIC,
                },
                {
                    cidrMask: 22,
                    name: 'private',
                    subnetType: ec2.SubnetType.PRIVATE_ISOLATED,
                },
            ],
        });
    }
}
