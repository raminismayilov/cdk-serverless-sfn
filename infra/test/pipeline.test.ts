import { App, Environment } from 'aws-cdk-lib';
import { Match, Template } from 'aws-cdk-lib/assertions';
import { PipelineStack } from '../lib/stacks/pipeline-stack';
import { ComputeStack } from '../lib/stacks/compute-stack';
import { BillingStack } from '../lib/stacks/billing-stack';

const testEnv = {
    account: '123456789012',
    region: 'us-east-1',
};

test('Pipeline Stack', () => {
    const app = new App();

    const stack = new PipelineStack(app, 'MyTestStack', {
        env: testEnv,
    });

    expect(Template.fromStack(stack).toJSON()).toMatchSnapshot();
});
