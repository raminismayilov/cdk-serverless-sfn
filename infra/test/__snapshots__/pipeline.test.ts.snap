// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pipeline Stack 1`] = `
Object {
  "Parameters": Object {
    "BootstrapVersion": Object {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": Object {
    "InfrastructureBuildProject8570F76B": Object {
      "Properties": Object {
        "Artifacts": Object {
          "Type": "CODEPIPELINE",
        },
        "Cache": Object {
          "Type": "NO_CACHE",
        },
        "EncryptionKey": "alias/aws/s3",
        "Environment": Object {
          "ComputeType": "BUILD_GENERAL1_SMALL",
          "Image": "aws/codebuild/standard:5.0",
          "ImagePullCredentialsType": "CODEBUILD",
          "PrivilegedMode": false,
          "Type": "LINUX_CONTAINER",
        },
        "ServiceRole": Object {
          "Fn::GetAtt": Array [
            "InfrastructureBuildProjectRole151242CA",
            "Arn",
          ],
        },
        "Source": Object {
          "BuildSpec": "infra/build-specs/build-spec.yml",
          "Type": "CODEPIPELINE",
        },
      },
      "Type": "AWS::CodeBuild::Project",
    },
    "InfrastructureBuildProjectRole151242CA": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "codebuild.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "InfrastructureBuildProjectRoleDefaultPolicy95860DE9": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "logs:CreateLogGroup",
                "logs:CreateLogStream",
                "logs:PutLogEvents",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":logs:us-east-1:123456789012:log-group:/aws/codebuild/",
                      Object {
                        "Ref": "InfrastructureBuildProject8570F76B",
                      },
                    ],
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":logs:us-east-1:123456789012:log-group:/aws/codebuild/",
                      Object {
                        "Ref": "InfrastructureBuildProject8570F76B",
                      },
                      ":*",
                    ],
                  ],
                },
              ],
            },
            Object {
              "Action": Array [
                "codebuild:CreateReportGroup",
                "codebuild:CreateReport",
                "codebuild:UpdateReport",
                "codebuild:BatchPutTestCases",
                "codebuild:BatchPutCodeCoverages",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:",
                    Object {
                      "Ref": "AWS::Partition",
                    },
                    ":codebuild:us-east-1:123456789012:report-group/",
                    Object {
                      "Ref": "InfrastructureBuildProject8570F76B",
                    },
                    "-*",
                  ],
                ],
              },
            },
            Object {
              "Action": Array [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
                "s3:DeleteObject*",
                "s3:PutObject",
                "s3:PutObjectLegalHold",
                "s3:PutObjectRetention",
                "s3:PutObjectTagging",
                "s3:PutObjectVersionTagging",
                "s3:Abort*",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "PipelineArtifactsBucket22248F97",
                    "Arn",
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      Object {
                        "Fn::GetAtt": Array [
                          "PipelineArtifactsBucket22248F97",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "InfrastructureBuildProjectRoleDefaultPolicy95860DE9",
        "Roles": Array [
          Object {
            "Ref": "InfrastructureBuildProjectRole151242CA",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "PipelineArtifactsBucket22248F97": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "BucketEncryption": Object {
          "ServerSideEncryptionConfiguration": Array [
            Object {
              "ServerSideEncryptionByDefault": Object {
                "SSEAlgorithm": "aws:kms",
              },
            },
          ],
        },
        "PublicAccessBlockConfiguration": Object {
          "BlockPublicAcls": true,
          "BlockPublicPolicy": true,
          "IgnorePublicAcls": true,
          "RestrictPublicBuckets": true,
        },
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "PipelineArtifactsBucketPolicyD4F9712A": Object {
      "Properties": Object {
        "Bucket": Object {
          "Ref": "PipelineArtifactsBucket22248F97",
        },
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "s3:*",
              "Condition": Object {
                "Bool": Object {
                  "aws:SecureTransport": "false",
                },
              },
              "Effect": "Deny",
              "Principal": Object {
                "AWS": "*",
              },
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "PipelineArtifactsBucket22248F97",
                    "Arn",
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      Object {
                        "Fn::GetAtt": Array [
                          "PipelineArtifactsBucket22248F97",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::S3::BucketPolicy",
    },
    "PipelineBuildPipelineBuildCodePipelineActionRoleB09C1912": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "AWS": Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::123456789012:root",
                    ],
                  ],
                },
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "PipelineBuildPipelineBuildCodePipelineActionRoleDefaultPolicy250B3BF8": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "codebuild:BatchGetBuilds",
                "codebuild:StartBuild",
                "codebuild:StopBuild",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "InfrastructureBuildProject8570F76B",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "PipelineBuildPipelineBuildCodePipelineActionRoleDefaultPolicy250B3BF8",
        "Roles": Array [
          Object {
            "Ref": "PipelineBuildPipelineBuildCodePipelineActionRoleB09C1912",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "PipelineC660917D": Object {
      "DependsOn": Array [
        "PipelineRoleDefaultPolicyC7A05455",
        "PipelineRoleD68726F7",
      ],
      "Properties": Object {
        "ArtifactStore": Object {
          "Location": Object {
            "Ref": "PipelineArtifactsBucket22248F97",
          },
          "Type": "S3",
        },
        "Name": "MyPipeline",
        "RestartExecutionOnUpdate": true,
        "RoleArn": Object {
          "Fn::GetAtt": Array [
            "PipelineRoleD68726F7",
            "Arn",
          ],
        },
        "Stages": Array [
          Object {
            "Actions": Array [
              Object {
                "ActionTypeId": Object {
                  "Category": "Source",
                  "Owner": "ThirdParty",
                  "Provider": "GitHub",
                  "Version": "1",
                },
                "Configuration": Object {
                  "Branch": "master",
                  "OAuthToken": "{{resolve:secretsmanager:github-token:SecretString:::}}",
                  "Owner": "raminismayilov",
                  "PollForSourceChanges": false,
                  "Repo": "cdk-serverless-sfn",
                },
                "Name": "Source",
                "OutputArtifacts": Array [
                  Object {
                    "Name": "SourceOutput",
                  },
                ],
                "RunOrder": 1,
              },
            ],
            "Name": "Source",
          },
          Object {
            "Actions": Array [
              Object {
                "ActionTypeId": Object {
                  "Category": "Build",
                  "Owner": "AWS",
                  "Provider": "CodeBuild",
                  "Version": "1",
                },
                "Configuration": Object {
                  "ProjectName": Object {
                    "Ref": "InfrastructureBuildProject8570F76B",
                  },
                },
                "InputArtifacts": Array [
                  Object {
                    "Name": "SourceOutput",
                  },
                ],
                "Name": "PipelineBuild",
                "OutputArtifacts": Array [
                  Object {
                    "Name": "BuildOutput",
                  },
                ],
                "RoleArn": Object {
                  "Fn::GetAtt": Array [
                    "PipelineBuildPipelineBuildCodePipelineActionRoleB09C1912",
                    "Arn",
                  ],
                },
                "RunOrder": 1,
              },
            ],
            "Name": "Build",
          },
          Object {
            "Actions": Array [
              Object {
                "ActionTypeId": Object {
                  "Category": "Deploy",
                  "Owner": "AWS",
                  "Provider": "CloudFormation",
                  "Version": "1",
                },
                "Configuration": Object {
                  "ActionMode": "CREATE_UPDATE",
                  "Capabilities": "CAPABILITY_NAMED_IAM",
                  "RoleArn": Object {
                    "Fn::GetAtt": Array [
                      "PipelinePipelineUpdatePipelineUpdateRole205CA4C4",
                      "Arn",
                    ],
                  },
                  "StackName": "PipelineStack",
                  "TemplatePath": "BuildOutput::PipelineStack.template.json",
                },
                "InputArtifacts": Array [
                  Object {
                    "Name": "BuildOutput",
                  },
                ],
                "Name": "PipelineUpdate",
                "RoleArn": Object {
                  "Fn::GetAtt": Array [
                    "PipelinePipelineUpdatePipelineUpdateCodePipelineActionRole1B522132",
                    "Arn",
                  ],
                },
                "RunOrder": 1,
              },
            ],
            "Name": "Pipeline_Update",
          },
        ],
      },
      "Type": "AWS::CodePipeline::Pipeline",
    },
    "PipelinePipelineUpdatePipelineUpdateCodePipelineActionRole1B522132": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "AWS": Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::123456789012:root",
                    ],
                  ],
                },
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "PipelinePipelineUpdatePipelineUpdateCodePipelineActionRoleDefaultPolicy08D5C704": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "iam:PassRole",
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "PipelinePipelineUpdatePipelineUpdateRole205CA4C4",
                  "Arn",
                ],
              },
            },
            Object {
              "Action": Array [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "PipelineArtifactsBucket22248F97",
                    "Arn",
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      Object {
                        "Fn::GetAtt": Array [
                          "PipelineArtifactsBucket22248F97",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
            Object {
              "Action": Array [
                "cloudformation:CreateStack",
                "cloudformation:DescribeStack*",
                "cloudformation:GetStackPolicy",
                "cloudformation:GetTemplate*",
                "cloudformation:SetStackPolicy",
                "cloudformation:UpdateStack",
                "cloudformation:ValidateTemplate",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:",
                    Object {
                      "Ref": "AWS::Partition",
                    },
                    ":cloudformation:us-east-1:123456789012:stack/PipelineStack/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "PipelinePipelineUpdatePipelineUpdateCodePipelineActionRoleDefaultPolicy08D5C704",
        "Roles": Array [
          Object {
            "Ref": "PipelinePipelineUpdatePipelineUpdateCodePipelineActionRole1B522132",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "PipelinePipelineUpdatePipelineUpdateRole205CA4C4": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "cloudformation.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "PipelinePipelineUpdatePipelineUpdateRoleDefaultPolicyE058B9A6": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "PipelineArtifactsBucket22248F97",
                    "Arn",
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      Object {
                        "Fn::GetAtt": Array [
                          "PipelineArtifactsBucket22248F97",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
            Object {
              "Action": "*",
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "PipelinePipelineUpdatePipelineUpdateRoleDefaultPolicyE058B9A6",
        "Roles": Array [
          Object {
            "Ref": "PipelinePipelineUpdatePipelineUpdateRole205CA4C4",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "PipelineRoleD68726F7": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "codepipeline.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "PipelineRoleDefaultPolicyC7A05455": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
                "s3:DeleteObject*",
                "s3:PutObject",
                "s3:PutObjectLegalHold",
                "s3:PutObjectRetention",
                "s3:PutObjectTagging",
                "s3:PutObjectVersionTagging",
                "s3:Abort*",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "PipelineArtifactsBucket22248F97",
                    "Arn",
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      Object {
                        "Fn::GetAtt": Array [
                          "PipelineArtifactsBucket22248F97",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "PipelineBuildPipelineBuildCodePipelineActionRoleB09C1912",
                  "Arn",
                ],
              },
            },
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "PipelinePipelineUpdatePipelineUpdateCodePipelineActionRole1B522132",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "PipelineRoleDefaultPolicyC7A05455",
        "Roles": Array [
          Object {
            "Ref": "PipelineRoleD68726F7",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "PipelineSourceWebhookResourceD1089B44": Object {
      "Properties": Object {
        "Authentication": "GITHUB_HMAC",
        "AuthenticationConfiguration": Object {
          "SecretToken": "{{resolve:secretsmanager:github-token:SecretString:::}}",
        },
        "Filters": Array [
          Object {
            "JsonPath": "$.ref",
            "MatchEquals": "refs/heads/{Branch}",
          },
        ],
        "RegisterWithThirdParty": true,
        "TargetAction": "Source",
        "TargetPipeline": Object {
          "Ref": "PipelineC660917D",
        },
        "TargetPipelineVersion": 1,
      },
      "Type": "AWS::CodePipeline::Webhook",
    },
  },
  "Rules": Object {
    "CheckBootstrapVersion": Object {
      "Assertions": Array [
        Object {
          "Assert": Object {
            "Fn::Not": Array [
              Object {
                "Fn::Contains": Array [
                  Array [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  Object {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
